core/src/Cuda/Kokkos_Cuda_Instance.cpp

```diff
@@ -684,6 +690,32 @@ Cuda::size_type *cuda_internal_scratch_unified(const Cuda &instance,
 
 namespace Kokkos {
 
+bool Cuda::is_running() const {
+  switch (m_space_instance->m_internal_status) {
+    case Impl::ExecutionSpaceStatus::complete: return false;
+    case Impl::ExecutionSpaceStatus::submitted:
+      KOKKOS_IMPL_CUDA_SAFE_CALL(cudaEventRecord(m_space_instance->m_last_event,
+                                                 m_space_instance->m_stream));
+      m_space_instance->m_internal_status = Impl::ExecutionSpaceStatus::running;
+      [[fallthrough]];
+    case Impl::ExecutionSpaceStatus::running: {
+      const cudaError_t query_result =
+          cudaEventQuery(m_space_instance->m_last_event);
+      if (query_result == cudaSuccess) {
```

Comment:
With `m_internal_status` being implemented as an atomic, we're not going to have data races resulting from accesses to it.

However, there's still a dangerous semantic race condition, in that another thread could submit a kernel launch here. If that happened, its kernel could be running, but its assignment of `m_internal_status = submitted` could be incorrectly overwritten by `complete`.